src/main/scala/firrtl/graph/EulerTour.scala

```diff
@@ -0,0 +1,224 @@
+// See LICENSE for license details.
+
+package firrtl.graph
+
+import scala.collection.mutable
+
+/** Euler Tour companion object */
+object EulerTour {
+  /** Create an Euler Tour of a `DiGraph[T]` */
+  def apply[T](diGraph: DiGraph[T], start: T): EulerTour[Seq[T]] = {
+    val r = mutable.Map[Seq[T], Int]()
+    val e = mutable.ArrayBuffer[Seq[T]]()
+    val h = mutable.ArrayBuffer[Int]()
+
+    def tour(u: T, parent: Seq[T] = Seq(), height: Int = 0): Unit = {
+      val id = parent :+ u
+      if (!r.contains(id)) { r(id) = e.size }
+      e += id
+      h += height
+      diGraph.getEdges(id.last).toSeq.flatMap( v => {
```

Comment:
May I ask why the edges need to be converted to a Seq and why a flatMap? I think this should just be a foreach on the edges.

Reply:
This is likely just an inefficient approach. I'm usually testing things on the REPL to get the behavior that I want. Usually, I'm trying to `map` everything and don't fallback to the obvious `foreach`. That would simplify things and avoid the conversion cost.